### ISSUE TITLE
Suggestion: Player Address Banning & Bounding

### ISSUE THREAD

**Issue by Maitlans** [enjin/enjincraft#61]
I think these are important for server owners to moderate their server and protection against potential abuse. Please let me know what you think, thank you.

**Banning Addresses**: You have the ability to ban wallets so they can't be linked to any users on your server. This can be done to players whom perhaps purchased something and charged back or players who've abused some sort of bug or just in general banning from playing.

**Limiting Addresses**: Option to bound addresses to their first linked Minecraft accounts, past this they are not allowed to link to any other Minecraft UUID, the reason for this would be to prevent abuse, players would be able to share their in-game items/permissions freely with others by giving them their Link code and approving it and it's secure as the other user has no way of stealing anything because it's all done by Requests.  Alternatively an option to have a cooldown of when to link again to another account. Also users have the option to contact server owner for them to reset the bounding/cooldown if they have a good enough reason.

**Comment by Favorlock** [comment]
I don't think this will be possible. Linking is handled in the app wallet, so we ultimately can't prevent linking an address.

**Comment by Maitlans** [comment]
> I don't think this will be possible. Linking is handled in the app wallet, so we ultimately can't prevent linking an address.

Do you think it would be possible to instantly reject the wallet after it links or just not allow it to be given any permissions or ability to use items, just blocked from all EnjinCraft features? This could be a bypass to that for Banning & Limiting addresses to 1 per user.

**Comment by Favorlock** [comment]
It'd be possible to forcibly unlink the identity from the wallet address, but I don't really feel like banning addresses is the right way to handle things. Addresses are easily created, so a person could easily get around this by simply creating a new address.

**Comment by Maitlans** [comment]
The first suggestion is to discourage rule breakers at least having them go through the trouble of sending all their assets to another wallet, rule breakers come in all forms and I think at least having a semi defense against them would be good, and the second one is to entirely prevent wallet sharing abuse which can consist of sharing it with friends and alt accounts, there's sites which generate you a Minecraft login cookie for free to random accounts.  Wallet sharing can be really disruptive to a server & how they function as well as exploits can come of it. EnjinCraft currently doesn't have moderation tools so I'm suggesting these two which I think can have a long term positive effect.

**Comment by Favorlock** [comment]
Wallet sharing shouldn't be an issue because only one player can be linked to an address at any given moment. Even if two people were to share an address they couldn't simultaneously play on the server and access those assets as the plugin detects unlink events.

**Comment by Maitlans** [comment]
Hey, I want to share some examples of issues that can cause, I really appreciate you taking the time to discuss with me.

Bypassing Enjin transaction fees.

Bypassing the need to sell on the Enjin Marketplace when they can just have the person send them funds and they share their wallet with them to give them access to all their items and permisisons for a set amount of time.

Let's say a player is playing a dungeon and they obtain a collectible which is a treasure map and gives them the ability to dig in this certain spot for a treasure chest & it has 1 Enjin transaction fee, the player can go ahead and wallet share to all his friends so they too can access that treasure without them actually owning the token & the server owner doesn't get that 1 Enjin fee. Everything having to do with permissions this can be quests, blocked off areas, treasure maps and other ideas that can be done via permissions can be shared. This can be done with weapons & armor too.

Sharing isn't the issue the whole problem is every time a wallet is shared that's 1 less Enjin fee per item the developer doesn't get to help fund their game/server. I understand this is an issue every game will face but at least on EnjinCraft's side it'd have an option to have server owners choose what they want to do and prevent it. Right now I understand why it wouldn't seem like such a big issue right now but it could become the norm for players to do and I think it's one that can have negative impacts on profitability and longevity for projects. Thank you for replying I think it's good to have these types of discussions & bring up potential issues that can arise before they even happen.
